rundeckapp/grails-app/services/rundeck/services/JobLifecycleComponentService.groovy

```diff
@@ -158,88 +188,77 @@ class JobLifecyclePluginService implements ProjectConfigurable {
      *
      * @param event job event
      * @param eventType type of event
-     * @param plugins list of NamedJobLifecyclePlugin
+     * @param components list of NamedJobLifecycleComponent
      * @return JobEventStatus response from plugin implementation
      */
-    JobLifecycleStatus handleEvent(def event, EventType eventType, List<NamedJobLifecyclePlugin> plugins)
-            throws JobLifecyclePluginException {
-        if (!plugins) {
+    JobLifecycleStatus handleEvent(def receivedEvent, EventType eventType, List<NamedJobLifecycleComponent> components)
+            throws JobLifecycleComponentException {
+        
+        if (!components) {
             return null
         }
-        def errors = [:]
-        def results = [:]
-        Exception firstErr
-        JobLifecycleStatus prevResult = null
-        def prevEvent = event
+        
+        Map<String, Exception> errors = [:]
         boolean success = true
-        for (NamedJobLifecyclePlugin plugin : plugins) {
+        boolean useNewOptionValues = false
+        boolean useNewExecutionMetadata = false
+        
+        def currentEvent = receivedEvent
+        
+        for (NamedJobLifecycleComponent component : components) {
             try {
+                JobLifecycleStatus result = handleEventForPlugin(eventType, component, currentEvent)
 
-                def curEvent = mergeEvent(prevResult, prevEvent)
-                JobLifecycleStatus result = handleEventForPlugin(eventType, plugin, curEvent)
                 if (result != null && !result.successful) {
-                    success = false
-                    if (result != null && result.getErrorMessage() != null && !result.getErrorMessage().trim().isEmpty()) {
-                        throw new JobLifecyclePluginException(result.getErrorMessage())
-                    } else {
-                        throw new JobLifecyclePluginException(
-                                "Job Lifecycle Plugin: $plugin.name: result for $eventType: was unsuccessful: no message"
-                        )
-                    }
-
+                    throw new JobLifecycleComponentException("Error dispatcing event [${eventType}] for component [${component.name}]: " + result?.errorMessage)
                 }
-                if (result != null && result.isUseNewValues()) {
-                    results[plugin.name] = result
-                    prevResult = result
-                }
-                prevEvent = curEvent
-            } catch (Throwable e) {
+                
+                useNewOptionValues = result?.isUseNewValues() || useNewOptionValues
+                useNewExecutionMetadata = result?.isUseNewMetadata() || useNewExecutionMetadata
+                currentEvent = mergeEvent(result, currentEvent)
+                
+            } catch (Exception e) {
```

Comment:
Catching throwable is generally considered a bad practice, as it will catch both exceptions (checked and unchecked) but also VM Errors (like OutOfMemoryError, or StackOverflowError). 

Unless you actually want to program certain behavior for Errors (like having a routine that frees memory when a OOM is encountered to allow application to continue), the recommendation is to just leave them to the JVM and let the application to crash.

There's plenty of literature about this. an example: https://www.baeldung.com/java-catch-throwable-bad-practice